core/coreunix/add.go

```diff
@@ -19,22 +27,146 @@ import (
 
 var log = eventlog.Logger("coreunix")
 
+// how many bytes of progress to wait before sending a progress update message
+const progressReaderIncrement = 1024 * 256
+
+type Link struct {
+	Name, Hash string
+	Size       uint64
+}
+
+type Object struct {
+	Hash  string
+	Links []Link
+}
+
+type hiddenFileError struct {
+	fileName string
+}
+
+func (e *hiddenFileError) Error() string {
+	return fmt.Sprintf("%s is a hidden file", e.fileName)
+}
+
+type ignoreFileError struct {
+	fileName string
+}
+
+func (e *ignoreFileError) Error() string {
+	return fmt.Sprintf("%s is an ignored file", e.fileName)
+}
+
+type AddedObject struct {
+	Name  string
+	Hash  string `json:",omitempty"`
+	Bytes int64  `json:",omitempty"`
+}
+
+func NewAdder(ctx context.Context, n *core.IpfsNode, out chan interface{}) *Adder {
+	e := dagutils.NewDagEditor(NewMemoryDagService(), newDirNode())
+	return &Adder{
+		ctx:      ctx,
+		node:     n,
+		editor:   e,
+		out:      out,
+		Progress: false,
+		Hidden:   true,
+		Pin:      true,
+		Trickle:  false,
+		Wrap:     false,
+		Chunker:  "",
+	}
+}
+
+// Internal structure for holding the switches passed to the `add` call
+type Adder struct {
+	ctx      context.Context
+	node     *core.IpfsNode
+	editor   *dagutils.Editor
+	out      chan interface{}
+	Progress bool
+	Hidden   bool
+	Pin      bool
+	Trickle  bool
+	Wrap     bool
+	Chunker  string
+	root     *merkledag.Node
+}
+
+// Perform the actual add & pin locally, outputting results to reader
+func (params Adder) add(reader io.Reader) (*merkledag.Node, error) {
+	chnk, err := chunk.FromString(reader, params.Chunker)
+	if err != nil {
+		return nil, err
+	}
+
+	if params.Trickle {
+		return importer.BuildTrickleDagFromReader(
+			params.node.DAG,
+			chnk,
+		)
+	}
+	return importer.BuildDagFromReader(
+		params.node.DAG,
+		chnk,
+	)
+}
+
+func (params *Adder) RootNode() (*merkledag.Node, error) {
+	// for memoizing
+	if params.root != nil {
+		return params.root, nil
+	}
+
+	root := params.editor.GetNode()
+
+	// if not wrapping, AND one root file, use that hash as root.
+	if !params.Wrap && len(root.Links) == 1 {
+		var err error
+		root, err = root.Links[0].GetNode(params.ctx, params.editor.GetDagService())
+		params.root = root
+		// no need to output, as we've already done so.
+		return root, err
+	}
+
+	// otherwise need to output, as we have not.
+	err := outputDagnode(params.out, "", root)
+	params.root = root
+	return root, err
+}
+
+func (params *Adder) PinRoot() error {
+	root, err := params.RootNode()
+	if err != nil {
+		return err
+	}
+
+	rnk, err := root.Key()
+	if err != nil {
+		return err
+	}
+
+	params.node.Pinning.PinWithMode(rnk, pin.Recursive)
+	return params.node.Pinning.Flush()
+}
+
+func (params *Adder) WriteOutputToDAG() error {
```

Comment:
The reason i kept the dagservice as the argument was so that we could use a temporary dagservice for some operation, and then write the final results to the actual dagservice when we were done, skipping a ton of intermediate stuff along the way